File: coding-interview-sorting-algorithms-pro/src/main/java/com/example/sorting/utils/ArrayUtils.java
```java
package com.example.sorting.utils;

import java.util.Arrays;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Utility class for array operations, helpful for testing and debugging sorting algorithms.
 */
public class ArrayUtils {

    /**
     * Prints an integer array to the console.
     *
     * @param arr The array to print.
     */
    public static void printArray(int[] arr) {
        System.out.println(Arrays.toString(arr));
    }

    /**
     * Swaps two elements in an array.
     *
     * @param arr The array.
     * @param i   Index of the first element.
     * @param j   Index of the second element.
     */
    public static void swap(int[] arr, int i, int j) {
        if (i == j) { // Optimization: no need to swap if indices are the same
            return;
        }
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    /**
     * Checks if an array is sorted in ascending order.
     *
     * @param arr The array to check.
     * @return true if the array is sorted, false otherwise.
     */
    public static boolean isSorted(int[] arr) {
        if (arr == null || arr.length <= 1) {
            return true; // Empty or single-element arrays are considered sorted
        }
        for (int i = 0; i < arr.length - 1; i++) {
            if (arr[i] > arr[i + 1]) {
                return false;
            }
        }
        return true;
    }

    /**
     * Generates a random integer array of a specified size.
     * Elements are within a default range of [0, 1000).
     *
     * @param size The desired size of the array.
     * @return A randomly generated integer array.
     */
    public static int[] generateRandomArray(int size) {
        return generateRandomArray(size, 0, 1000);
    }

    /**
     * Generates a random integer array of a specified size with elements within a given range.
     *
     * @param size The desired size of the array.
     * @param min  The minimum possible value for an element (inclusive).
     * @param max  The maximum possible value for an element (exclusive).
     * @return A randomly generated integer array.
     */
    public static int[] generateRandomArray(int size, int min, int max) {
        if (size < 0) {
            throw new IllegalArgumentException("Array size cannot be negative.");
        }
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = ThreadLocalRandom.current().nextInt(min, max);
        }
        return arr;
    }

    /**
     * Creates a deep copy of an integer array.
     *
     * @param original The array to copy.
     * @return A new array containing the same elements as the original.
     */
    public static int[] copyArray(int[] original) {
        if (original == null) {
            return null;
        }
        return Arrays.copyOf(original, original.length);
    }

    /**
     * Generates an array that is mostly sorted but with a few elements out of place.
     *
     * @param size The desired size of the array.
     * @param numDisruptions The number of swaps to make to disrupt the sorted order.
     * @return A mostly sorted array.
     */
    public static int[] generateMostlySortedArray(int size, int numDisruptions) {
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = i;
        }

        ThreadLocalRandom random = ThreadLocalRandom.current();
        for (int i = 0; i < numDisruptions; i++) {
            int idx1 = random.nextInt(size);
            int idx2 = random.nextInt(size);
            swap(arr, idx1, idx2);
        }
        return arr;
    }

    /**
     * Generates an array with many duplicate elements.
     *
     * @param size The desired size of the array.
     * @param uniqueElements The number of unique elements to use (must be less than or equal to size).
     * @return An array with duplicate elements.
     */
    public static int[] generateArrayWithDuplicates(int size, int uniqueElements) {
        if (uniqueElements <= 0 || uniqueElements > size) {
            throw new IllegalArgumentException("uniqueElements must be positive and less than or equal to size.");
        }
        int[] arr = new int[size];
        ThreadLocalRandom random = ThreadLocalRandom.current();
        for (int i = 0; i < size; i++) {
            arr[i] = random.nextInt(uniqueElements); // Values will be 0 to uniqueElements-1
        }
        return arr;
    }
}
```

File: data-structures-sorting-algorithms-advanced/src/utils/arrayUtils.js
```javascript
/**
 * @fileoverview Utility functions for array manipulation.
 */

/**
 * Swaps two elements in an array.
 * @param {Array<number>} arr The array to modify.
 * @param {number} i The index of the first element.
 * @param {number} j The index of the second element.
 * @returns {void} The array is modified in-place.
 */
function swap(arr, i, j) {
  if (i < 0 || i >= arr.length || j < 0 || j >= arr.length) {
    throw new Error('Invalid indices for swap operation.');
  }
  const temp = arr[i];
  arr[i] = arr[j];
  arr[j] = temp;
}

/**
 * Generates an array of random integers.
 * @param {number} size The desired size of the array.
 * @param {number} min The minimum value for elements (inclusive).
 * @param {number} max The maximum value for elements (inclusive).
 * @returns {Array<number>} A new array filled with random integers.
 */
function generateRandomArray(size, min = 0, max = 1000) {
  if (size < 0) {
    throw new Error('Array size cannot be negative.');
  }
  if (min > max) {
    throw new Error('Min value cannot be greater than max value.');
  }
  const arr = new Array(size);
  for (let i = 0; i < size; i++) {
    arr[i] = Math.floor(Math.random() * (max - min + 1)) + min;
  }
  return arr;
}

/**
 * Checks if an array is sorted in ascending or descending order.
 * @param {Array<number>} arr The array to check.
 * @param {boolean} [ascending=true] True for ascending order, false for descending.
 * @returns {boolean} True if the array is sorted as specified, false otherwise.
 */
function isSorted(arr, ascending = true) {
  if (!Array.isArray(arr)) {
    throw new Error('Input must be an array.');
  }
  if (arr.length <= 1) {
    return true; // An empty or single-element array is always sorted.
  }

  for (let i = 0; i < arr.length - 1; i++) {
    if (ascending) {
      if (arr[i] > arr[i + 1]) {
        return false;
      }
    } else {
      if (arr[i] < arr[i + 1]) {
        return false;
      }
    }
  }
  return true;
}

module.exports = {
  swap,
  generateRandomArray,
  isSorted,
};
```